### PR TITLE
`ocamlc -g1` : generate smaller, less complete bytecode debugging information 

### PR DESCRIPTION
**Pull request by xavierleroy**
Debugging information generated by `ocamlc -g` is quite large because it includes full typing and local variable information -- everything that `ocamldebug` needs.  If only stack backtraces are needed, smaller information would suffice, and result in smaller bytecode executables.

The first commit in this PR adds options `-g0`, `-g1`, `-g2` to `ocamlc` and `ocamlopt`.
    
* `-g0` is synonymous with `-no-g` and means "no debugging information at all".
    
* `-g1` means "generate just enough debug information for stack backtraces".
    
*  `-g2` means "generate full debug information".  Supported only by `ocamlc`.
    
`-g` now means "generate as much debug info as you can".   It is synonymous with `-g2` for bytecode and `-g1` for native-code.

The other commits implement `-g1` mode for bytecode compilation and linking.  It simply removes all typing informations and informations about local variables from the debugging events.  This does not affect the generation of stack backtraces, and some debugging with ocamldebug is still possible (but no printing of values, of course).
    
The `-g1` option takes effect when compiling to a .cmo but also when linking to a bytecode executable.  In particular, it is possible to compile a library with `-g`/`-g2` to get full debugging info, yet link with `-g1`, if smaller debugging info is desired.

Some measurements on the `ocamlopt` bytecode executable:
- Linked with `-g`: total size 26625263 bytes, including debug info 22636845 bytes
- Linked with `-g1`: total size 11723144 bytes (-56%), including debug info 7734726 (-66%).

### CODE REVIEW

**Comment by xavierleroy** [comment]
Note to reviewers: additional space savings could perhaps be obtained by having two kinds of debug info records, the full kind and the light kind, the latter having fewer fields, at the cost of more complex code in the bytecode linker and in ocamldebug.  But we're getting into "diminishing returns" territory.

**Comment by hhugo** [comment]
- There are 2 optims that are disabled when debug is on in `simpl.ml`.
I don't know if there are disabled to accommodate backtraces or ocamldebug workflow. I would be nice to document this.

- The new flags `g0`,`g1`,`g2` look reasonable at first, but I think one should take https://github.com/ocaml/ocaml/issues/6238 into account to make sure they can play nice together.

File: driver/main_args.ml [hunk 1]
```diff
@@ -140,7 +140,22 @@ let mk_g_opt f =
 
 let mk_no_g f =
   "-no-g", Arg.Unit f,
-  " Do not record debugging information (default)"
+  " Do not record debugging information (same as -g0) (default)"
+
+let mk_g2_byt f =
+  "-g2", Arg.Unit f, " Save debugging information (synonymous with -g)"
```
Comment (dra27) [inline driver/main_args.ml]: Possibly (cf. the description for `mk_no_g` above)
```suggestion
  "-g2", Arg.Unit f, " Save all debugging information (same as -g)"
```

**Review by dra27 — COMMENTED** [review]
Nice! (I agree with @hhugo that #6238 is worth consideration). Mechanism all looks fine to me - one possible tweak to argument description?

**Comment by xavierleroy** [comment]
Re: #6238, I think we could perform all the bytecode optimizations in -g1 mode, since the reason for disabling some of them is to keep source-level variables defined and inspectable by ocamldebug.  However, the bytecode generated with -g1 could still differ slightly from that generated with -g0, owing to the Levent / Kevent nodes that can prevent some instruction patterns from being recognized, despite our best efforts.

However, #6238 calls for something a bit different: full optimizations + full-but-possibly-inaccurate debugging info.  I'm not terribly interested in this combination myself, to be honest.

**Comment by lthls** [comment]
If I understand correctly, #6238 mostly calls for a flag that behaves like `-g` except for the code in `Cmm_helpers.raise_prim`. For full compatibility with the non-debug code, it should also enable the builtin methods optimisation in `Translclass` and the sharing of bound error sites in `Emit`, but it's probably less important.

But while it is reasonably easy to write the code for this feature, in terms of flag names it goes in the opposite direction from this PR: the "light" mode would enable debug symbols but not backtraces, while for bytecode it's the reverse.
One possibility would be to add a new flag `-gopt` for `ocamlopt` only, which triggers this mode. If it's acceptable from an UI point of view, I can submit a PR for that.

**Comment by edwintorok** [comment]
Thanks for the '-g1', I think that stacktrace info is too invaluable to be left out from final, production deployments.
Is the '-g2' vs '-g1' only something that can be done at compile time, or can '-g2' be stripped into a '-g1' + additional file, i.e. something that rpm/deb split-debuginfo packages could take advantage of (or rather copy the -g2 file into the debuginfo package, strip it into its -g1 form and install that in the binary package).

**Comment by hhugo** [comment]
Jsoo uses debug info to preserve variable names.  I've always been annoyed that `-g` disables some optimization. In particular, the elimination of local refs. I wonder if we could find a place for such mode. For instance, having `-g2` preserve optimizations and introduce `-g3` as `-g2` + some optimization disabled.

**Comment by xavierleroy** [comment]
I'm not sure I'm going to pursue this PR, as the space savings may not be enough to justify the additional complexity.

Re: debugging vs optimization, I now think that they should be handled by two different flags,
* `-g0` / `-g1` / `-g2` / `-g` to select the desired  amount of debugging info 
* `-O0` / `-O1` / `-O` to select the desired amount of optimizations.

We can have reasonable defaults, like `-O1` for `-g0` and `-g1` and `-O0` for `-g2`, but JSOO users could then use `-g2 -O1` and be happy.

**Comment by hhugo** [comment]
> Re: debugging vs optimization, I now think that they should be handled by two different flags,
> 
> * `-g0` / `-g1` / `-g2` / `-g` to select the desired  amount of debugging info
> * `-O0` / `-O1` / `-O` to select the desired amount of optimizations.
> 
> We can have reasonable defaults, like `-O1` for `-g0` and `-g1` and `-O0` for `-g2`, but JSOO users could then use `-g2 -O1` and be happy.

Sounds good

File: utils/clflags.ml [hunk 2]
```diff
@@ -51,7 +51,8 @@ and no_std_include = ref false          (* -nostdlib *)
 and no_cwd = ref false                  (* -nocwd *)
 and print_types = ref false             (* -i *)
 and make_archive = ref false            (* -a *)
-and debug = ref false                   (* -g *)
+and debug = ref false                   (* -g, -g2 *)
+and debug_light = ref false             (* -g1 *)
```
Comment (damiendoligez) [inline utils/clflags.ml:55]:
```suggestion
and debug = ref false                   (* -no-g, -g, -g0, -g1, -g2 *)
and debug_light = ref false             (* -g, -g1, -g2 *)
```

**Review by damiendoligez — DISMISSED** [review]
Looks good, although I'd mention (in the comments) all the options that can touch each flag.

**Comment by damiendoligez** [comment]
I think we can merge this PR first, then work on -O/-O1/-O2.

**Comment by damiendoligez** [comment]
I discussed this with @lthls and came to the conclusion that it makes more sense to have the option for ocamlopt called -g2 rather than -g1. This is because -g currently generates more than "just enough information for backtraces", it also adds some locations information that is used for debugging but not for backtraces.
Another possibility is to reserve g1/g2 to the bytecode where the distinction makes sense, and leave the native-code compiler with only -g/-no-g/-g0.

**Comment by xavierleroy** [comment]
Thanks for the review and feedback.  I propose to put this PR on hold while we're evaluating the benefits of compression #12006 .  Space savings from compression and  space savings from this PR are orthogonal, of course, but maybe the benefits of this PR are less convincing once compression is applied.  Also, as you note, the design of `-g0/-g1/-g2/-g` may need more work.

**Comment by xavierleroy** [comment]
Now that debugging information is compressed before storing, this PR is less useful and I don't feel like carrying it to completion.  Closing!

**Pull request closed without merging**